dll/cpl/sysdm/general.c

```diff
@@ -687,21 +698,47 @@ INT_PTR CALLBACK GeneralPageProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM
             SetWindowLongPtr(GetDlgItem(hwndDlg, IDC_ROSIMG), GWLP_WNDPROC, (LONG_PTR)RosImageProc);
             GetSystemInformation(hwndDlg);
             GetSystemVersion(hwndDlg);
-            GetSystemUptime(hwndDlg);
+            InitSystemUptime(hwndDlg);
             break;
+        }
 
         case WM_DESTROY:
+        {
+            KillTimer(hwndDlg, ID_SYSUPTIME_UPDATE_TIMER);
             HeapFree(GetProcessHeap(), 0, pImgInfo);
+
+            if (hKernel32Vista)
+            {
+                FreeLibrary(hKernel32Vista);
+            }
+
             break;
+        }
+
+        case WM_TIMER:
+        {
+            if (wParam == ID_SYSUPTIME_UPDATE_TIMER)
+            {
+                /* Update system uptime */
+                GetSystemUptime(hwndDlg);
+
+                /* Reset timeout */
+                SetTimer(hwndDlg, ID_SYSUPTIME_UPDATE_TIMER, 1000 - (RosUptime % 1000), NULL);
```

Comment:
Is the reset necessary? What about just letting the `WM_TIMER` being generated whenever the window is ready to receive the message (the timer is not destroyed automatically by default), and your GetSystemUptime function always displays the correct uptime, since it comes from the system values?

Reply:
> Or was it done so as to try to get the timer to trigger "almost exactly" as when the uptime should change?

Yes, the purpose is to get the correct update speed.
The idea is based on what's in timedate.cpl.
I tried hardcoding the speed to 1000 and the result was incorrect.